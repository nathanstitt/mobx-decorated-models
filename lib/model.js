class Box extends BaseModel {
    static attributes = {
        name: null,
        x: 1,
        y: 1,
        someAttribute: 'default value',
    }
    @computed get area() {
        console.log("COMPU");
        return this.x * this.y;
    }
}
