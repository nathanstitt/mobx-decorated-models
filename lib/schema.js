import { observable } from 'mobx';

export default function (model) {
    if (model.$schema) {
        return model.$schema;
    }
    const schema = observable.shallowMap();
    Object.defineProperty(model, '$schema', {
        enumerable: false,
        writable: false,
        configurable: true,
        value: schema,
    });
    return schema;
}
