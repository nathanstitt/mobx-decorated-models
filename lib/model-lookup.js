const ModelsMap = new Map();

function defaultModelLookup({ property, model = property }) {
    if (typeof model === 'function') { return model; }
    return ModelsMap[model] || ModelsMap[property];
}
let modelLookup = defaultModelLookup;

const defaultRememberModel = (model) => {
    ModelsMap[model.identifiedBy || model.name] = model;
};
let rememberModel = defaultRememberModel;

export function registerModel(model) {
    rememberModel(model);
}

export function findModel(model, property) {
    return modelLookup({ model, property });
}

export function rememberModelUsing(fn) {
    rememberModel = fn || defaultRememberModel;
}

export function lookupModelUsing(fn) {
    modelLookup = fn || defaultModelLookup;
}
