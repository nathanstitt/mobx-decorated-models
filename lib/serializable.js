export function isSerializable(model, property) {
    return !!(!model.$nonSerializable || !model.$nonSerializable[property]);
}

export function markNonserializable(model, property) {
    model.$nonSerializable = model.$nonSerializable || {};
    model.$nonSerializable[property] = true;
}
