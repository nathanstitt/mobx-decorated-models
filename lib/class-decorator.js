import {
    serialize, deserialize, createModelSchema, primitive, list, object,
    identifier, getDefaultModelSchema, update,
} from 'serializr';

import { registerModel, findModel } from './model-lookup';
import { isSerializable, markNonserializable } from './serializable';
import getSchema from './schema';

const PendingLookups = [];

function objectSerializer() {
    return {
        serializer(obj, prop, parent) {
            return isSerializable(parent, prop) ? obj : undefined;
        },
        deserializer(json, cb) {
            cb(null, json);
        },
    };
}

function addReference(parentModel, propName, options, cb) {
    const model = findModel(options.model, propName);
    if (model) {
        getDefaultModelSchema(parentModel).props[propName] = cb(model, options, propName);
    } else {
        getDefaultModelSchema(parentModel).props[propName] = objectSerializer();
        PendingLookups.push({ parentModel, propName, options, cb });
    }
}

function getSerializer(options, defaultSerializer) {
    let serializer;
    if (options.type === 'object') {
        serializer = objectSerializer;
    } else if (options.type === 'array') {
        serializer = list;
    } else {
        serializer = defaultSerializer;
    }
    const fns = serializer();
    if (options.writeOnly) {
        return {
            serializer: () => undefined,
            deserializer: fns.deserializer,
        };
    }
    return fns;
}

const Fields = {
    identifier: options => getSerializer(options, identifier),
    field: options => getSerializer(options, primitive),
    session: options =>
        getSerializer(Object.assign({}, options, { writeOnly: true }), primitive),
};

const Associations = {
    hasMany(modelKlass, options) {
        const defaultSerializer = list(object(modelKlass));
        return Object.assign(defaultSerializer, options, {
            model: modelKlass,
        });
    },
    belongsTo(modelKlass, options, propName) {
        const defaultSerializer = object(getDefaultModelSchema(modelKlass));
        return Object.assign(options, {
            model: modelKlass,
            deserializer(value, cb, context) {
                defaultSerializer.deserializer(value, (err, model) => {
                    if (!err && options.inverseOf) {
                        markNonserializable(model, options.inverseOf);
                        model[`${options.inverseOf}_association_name`] = propName;
                        model[options.inverseOf] = context.target;
                    }
                    cb(err, model);
                });
            },
            serializer(belongsTo, name, parent) {
                if (!isSerializable(parent, name)) { return undefined; }
                return defaultSerializer.serializer(belongsTo);
            },
        });
    },
};


const MixedInInstanceMethods = {
    serialize() {
        const schema = getDefaultModelSchema(this.constructor);
        return serialize(schema, this);
    },
    update(json, callback) {
        return update(getDefaultModelSchema(this.constructor), this, json, callback);
    },
};

const MixedInClassMethods = {
    deserialize(json, callback) {
        return deserialize(getDefaultModelSchema(this), json, callback);
    },
};

function decorateModel(model) {
    Object.assign(model.prototype, MixedInInstanceMethods);
    Object.assign(model, MixedInClassMethods);

    const schema = getSchema(model);
    registerModel(model);
    const serializeSchema = {};
    schema.forEach(({ type, options }, name) => {
        if (Fields[type]) {
            serializeSchema[name] = Fields[type](options, name);
        }
    });
    createModelSchema(model, serializeSchema);
    schema.forEach(({ type, options }, name) => {
        if (Associations[type]) {
            addReference(model, name, options, Associations[type]);
        }
    });
    for (let i = PendingLookups.length - 1; i >= 0; i -= 1) {
        const { parentModel, propName, options, cb } = PendingLookups[i];
        const referencedModel = findModel(options.model, propName);
        if (referencedModel) {
            const parentModelSchema = getDefaultModelSchema(parentModel);
            parentModelSchema.props[propName] = cb(referencedModel, options, propName);
            PendingLookups.splice(i, 1);
        }
    }
}

export function modelDecorator(modelOrIdentifier) {
    if (typeof modelOrIdentifier === 'function') {
        modelOrIdentifier.identifiedBy = modelOrIdentifier.name;
        return decorateModel(modelOrIdentifier);
    }
    return (model) => {
        model.identifiedBy = modelOrIdentifier;
        return decorateModel(model);
    };
}

export function unresolvedAssociations() {
    return PendingLookups.map(({ parentModel: model, propName: property }) => (
        { model, property }
    ));
}
