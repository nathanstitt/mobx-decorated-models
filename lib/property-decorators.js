import { observable, intercept } from 'mobx';
import { getDefaultModelSchema } from 'serializr';

import { findModel } from './model-lookup';
import getModelSchema from './schema';
import { markNonserializable } from './serializable'

function addLazyInitializer(target, fn) {
    target.__mobxLazyInitializers.push(fn);
}

function setupModel({
    attrs, modelClass: Klass, array, defaultAttributes,
    inverseOf, parentModel, parentModelProp,
}) {
    if (!attrs || typeof attrs !== 'object') {
        return attrs;
    }
    if (defaultAttributes) {
        if (typeof defaultAttributes === 'function') {
            defaultAttributes = defaultAttributes.call(parentModel, array, parentModel);
        }
        Object.keys(defaultAttributes).forEach((key) => {
            if (!attrs[key]) {
                attrs[key] = defaultAttributes[key];
            }
        });
    }
    if (inverseOf) {
        if (parentModelProp) {
            attrs[`${inverseOf}_association_name`] = parentModelProp;
        }
        attrs[inverseOf] = parentModel;
    }
    const model = (Klass && !(attrs instanceof Klass)) ? new Klass(attrs) : attrs;
    if (inverseOf) {
        markNonserializable(model, inverseOf);
    }
    return model;
}

function onBelongsToSet(change,
                        { modelClass, defaultAttributes, inverseOf,
                          parentModel, parentModelProp }) {
    change.newValue = setupModel({
        attrs: change.newValue, modelClass,
        defaultAttributes, inverseOf, parentModel, parentModelProp,
    });
    return change;
}

function onHasManySet(change,
                      { modelClass, defaultAttributes, inverseOf,
                        parentModel, parentModelProp }) {
    if (change.type !== 'update' || !change.newValue) { return change; }
    const { newValue: array } = change;

    for (let i = 0; i < array.length; i += 1) {
        array[i] = setupModel({
            attrs: array[i], array, modelClass,
            defaultAttributes, inverseOf, parentModel, parentModelProp,
        });
    }

    if (parentModel[parentModelProp]) {
        parentModel[parentModelProp].replace(array);
        return null;
    } else {
        change.newValue = observable.array(array);
        return change;
    }
}

function onCollectionChangeInterceptor(
    { modelClass, model, defaults: defaultAttributes, inverseOf },
    parentModel, parentModelProp,
) {
    return (change) => {
        if (!change.newValue) {
            change.newValue = {};
        }
        if (!modelClass) {
            modelClass = findModel(model, parentModelProp);
        }
        if (change.type === 'splice') {
            for (let i = 0; i < change.added.length; i += 1) {
                change.added[i] = setupModel({
                    attrs: change.added[i], array: change.object, modelClass,
                    defaultAttributes, inverseOf, parentModel, parentModelProp });
            }
        } else if (change.type === 'update') {
            change.newValue = setupModel({
                attrs: change.newValue, array: change.object, modelClass,
                defaultAttributes, inverseOf, parentModel, parentModelProp });
        }
        return change;
    };
}

function buildCollection(options, parentModel, parentModelProp) {
    const ary = observable.array([]);
    if (options.classId || options.model) {
        ary.intercept(onCollectionChangeInterceptor(options, parentModel, parentModelProp));
    }
    return ary;
}

const Initializers = {
    object: () => observable.object({}),
    array: buildCollection,
};

const TypeInitializers = {
    hasMany: buildCollection,
};

function getInitializer(type, options, propName) {
    const fn = TypeInitializers[type] || Initializers[options.type];
    if (!fn) { return undefined; }
    return function () { // eslint-disable-line func-names
        return fn(options, this, propName);
    };
}

function addAttribute(type, target, property, descriptor, options = {}) {
    getModelSchema(target.constructor).set(property, { name: property, type, options });
    const initializer = getInitializer(type, options, property);
    if (initializer) {
        descriptor.initializer = initializer;
    }
    const definition = observable(target, property, descriptor);
    return definition;
}

function buildAttributeDecorator(type, args, attributeAddFn = addAttribute) {
    if ((typeof args[0] === 'object') && args.length === 1) {
        return (target, property, descriptor) =>
            attributeAddFn(type, target, property, descriptor, args[0]);
    }
    return attributeAddFn(type, ...args, {});
}

function interceptingDecorator(interceptingFn) {
    return (type, target, property, descriptor, options = {}) => {
        const decorator = addAttribute(type, target, property, descriptor, options);
        addLazyInitializer(target, (model) => {
            const schema = getDefaultModelSchema(target);
            if (schema && schema.props[property]) {
                const schemaProps = schema.props[property];
                intercept(model, property, (change => interceptingFn(change, {
                    inverseOf:         schemaProps.inverseOf,
                    modelClass:        schemaProps.model,
                    parentModel:       model,
                    parentModelProp:   property,
                    defaultAttributes: schemaProps.defaults,
                })));
            }
        });
        return decorator;
    };
}


const hasMany = (...args) => buildAttributeDecorator(
    'hasMany', args, interceptingDecorator(onHasManySet),
);

const belongsTo = (...args) => buildAttributeDecorator(
    'belongsTo', args, interceptingDecorator(onBelongsToSet),
);

const identifier = (...args) => buildAttributeDecorator('identifier', args);
const field = (...args) => buildAttributeDecorator('field', args);
const session = (...args) => buildAttributeDecorator('session', args);

export { field, session, belongsTo, hasMany, identifier, buildCollection };
