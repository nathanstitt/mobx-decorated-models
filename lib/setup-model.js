import { markNonserializable } from './serializable';

export default function setupModel({
    attrs, modelClass: Klass, array, defaultAttributes,
    inverseOf, parentModel, parentModelProp,
}) {
    if (!attrs) { return attrs; }
    const isObject = ('object' === typeof attrs);
    const canWrite = Object.isExtensible(attrs);

    if (isObject && canWrite) {
        if (defaultAttributes) {
            if ('function' === typeof defaultAttributes) {
                defaultAttributes = defaultAttributes.call(parentModel, array, parentModel);
            }
            Object.keys(defaultAttributes).forEach((key) => {
                if (!attrs[key]) {
                    attrs[key] = defaultAttributes[key];
                }
            });
        }

        if (inverseOf) {
            attrs[inverseOf] = parentModel;
            if (parentModelProp) {
                attrs[`${inverseOf}_association_name`] = parentModelProp;
            }
        }
    }

    const model = (Klass && 'function' === typeof Klass && !(attrs instanceof Klass)) ?
        new Klass(attrs) : attrs;

    if (inverseOf) {
        markNonserializable(model, inverseOf);
    }

    return model;
}
