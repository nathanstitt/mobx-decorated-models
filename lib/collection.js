import { observable } from 'mobx';

import { findModel } from './model-lookup';
import { setupModel } from './setup-model';

function onCollectionChangeInterceptor(
    { modelClass, model, defaults: defaultAttributes, inverseOf },
    parentModel, parentModelProp,
) {
    return (change) => {
        if (!change.newValue) {
            change.newValue = {};
        }
        if (!modelClass) {
            modelClass = findModel(model, parentModelProp);
        }
        // console.log("ARY CHANGE");
        if (change.type === 'splice') {
            for (let i = 0; i < change.added.length; i += 1) {
                change.added[i] = setupModel({
                    attrs: change.added[i],
                    array: change.object,
                    modelClass,
                    defaultAttributes,
                    inverseOf,
                    parentModel,
                    parentModelProp });
            }
        } else if (change.type === 'update') {
            change.newValue = setupModel({
                attrs: change.newValue,
                array: change.object,
                modelClass,
                defaultAttributes,
                inverseOf,
                parentModel,
                parentModelProp });
        }
        return change;
    };
}


export default function createCollection(options, parentModel, parentModelProp) {
    const ary = observable.array([]);
    //console.log("CREATE COL", parentModelProp);
    if (options) {
        if (options.model) {
            ary.intercept(onCollectionChangeInterceptor(options, parentModel, parentModelProp));
        }
        if (options.extend) {
            if ('function' === typeof options.extend){
                options.extend(ary);
            } else {
                Object.assign(ary, options.extend);
            }
        }
    }

    return ary;
}
