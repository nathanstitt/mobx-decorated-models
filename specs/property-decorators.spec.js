import { autorun } from 'mobx';
import { Box, Container, Ship } from './test-models';
import { isSerializable } from '../lib/serializable';

describe('Property Decorators', () => {

    it('it doesn’t interfere with inheritance', () => {
        const container = new Container();
        expect(container.isCuboid).toEqual(true);
    });

    it('has attributes that are observable', () => {
        const box = new Box();
        box.width = 3;
        const spy = jest.fn();
        autorun(() => {
            spy(box.volume);
        });
        expect(spy).toHaveBeenCalledTimes(1);
        box.width = 3; // is already 3
        expect(spy).toHaveBeenCalledTimes(1);
        box.width = 8;
        expect(spy).toHaveBeenCalledTimes(2);
        box.temp = 'test, test'; // non existant field
        expect(spy).toHaveBeenCalledTimes(2);
    });

    it('sets attributes using provided values', () => {
        const box = new Box();
        box.update({ width: 3, depth: 12, height: 2, bad: true });
        expect(box.width).toEqual(3);
        expect(box.bad).toBeUndefined();
        expect(box.volume).toEqual(72);
    });

    it('sets defaults for hasMany', () => {
        const container = Container.deserialize({ color: 'blue' });
        container.boxes.push({});
        expect(container.boxes[0].color).toEqual('blue');
    });

    it('sets an inverse for hasMany', () => {
        const container = Container.deserialize({ id: 1, name: 'Bob', location: 'water' });
        container.boxes.push({});
        expect(container.boxes[0]).toBeInstanceOf(Box);
        expect(container.boxes[0].container).toBe(container);
        expect(container.boxes[0].container_association_name).toEqual('boxes');
    });

    it('finds model for belongsTo', () => {
        const box = Box.deserialize({ id: 1, watercraft: { name: 'Boaty' } });
        expect(box.watercraft).toBeInstanceOf(Ship);
    });

    it('sets an inverse for belongsTo', () => {
        const ship = Ship.deserialize({ name: 'HMS Mobx', box: { width: 42 } });
        expect(ship.box.depth).toEqual(1);
        expect(isSerializable(ship, 'box')).toBe(true);
        expect(isSerializable(ship.box, 'vessel')).toBe(false);
        expect(ship.serialize()).toEqual({
            name: 'HMS Mobx',
            box: { depth: 1, height: 1, metadata: {}, width: 42 },
        });
        expect(ship.box.vessel_association_name).toEqual('box');
    });

    it('merges both attributes and session props', () => {
        const box = Box.deserialize({ width: 3, color: 'red' });
        expect(box.color).toEqual('red');
        // no color
        expect(box.serialize()).toEqual({ depth: 1, height: 1, metadata: {}, width: 3 });
    });

    it('can observe associations', () => {
        const container = Container.deserialize({ id: 1, name: 'Bob', location: 'water' });
        const spy = jest.fn();
        autorun(() => {
            spy(container.boxes.length);
        });
        expect(container.areaInUse).toEqual(0);
        container.boxes.push({ id: 1, width: 8, depth: 12, height: 8 });
        container.boxes.push({ id: 2, width: 3, depth: 12, height: 4 });
        container.boxes[1].width = 4;
        expect(container.areaInUse).toEqual(960);
        expect(spy).toHaveBeenCalledTimes(3);
    });

    it('can set property to be object', () => {
        const box = Box.deserialize({
            width: 3, metadata: { barcode: 'Z12', color: 'black' },
        });
        expect(box.metadata).toEqual({ barcode: 'Z12', color: 'black' });
        expect(box.serialize()).toEqual({
            container: undefined, depth: 1, height: 1, id: undefined, width: 3,
            metadata: { barcode: 'Z12', color: 'black' },
        });
    });

    it('can set property to be array', () => {
        const tags = ['one', 'two', 'three'];
        const container = Container.deserialize({ id: 1, tags });
        expect(container.tags.slice()).toEqual(tags);
        expect(container.serialize()).toEqual({
            boxes: [], id: 1, location: undefined, name: undefined, tags,
        });
    });

    it('guards against setting a belongsTo', () => {
        const ship = Ship.deserialize({ name: 'HMS Glory' });
        ship.box = { width: 1, height: 1, depth: 1 };
        expect(ship.box).toBeInstanceOf(Box);
        expect(ship.box.vessel).toBe(ship);
    });

    it('passes on values that can’t be handled', () => {
        const ship = Ship.deserialize({ name: 'HMS Glory' });
        expect(ship.box).toBeUndefined();
        ship.box = undefined;
        expect(ship.box).toBeUndefined();
        ship.box = 1;
        expect(ship.box).toBe(1);
        const container = new Container();
        container.boxes = [{ id: 1 }];
        expect(container.boxes[0]).toBeInstanceOf(Box);
        container.boxes = undefined;
        expect(container.boxes).toBeUndefined();
        container.boxes = [{ id: 1 }, null];
        expect(container.boxes[0].id).toBe(1);
        expect(container.boxes[1]).toBeNull();
    });

    it('guards against setting a hasMany', () => {
        const container = Container.deserialize({ id: 1 });
        const originalBoxes = container.boxes;
        container.boxes = [
            { width: 1, height: 1, depth: 1 },
            { width: 2, height: 2, depth: 2 },
        ];
        expect(container.boxes).toBe(originalBoxes);
        expect(container.boxes[0]).toBeInstanceOf(Box);
        expect(container.boxes[1].container).toBe(container);
    });


});
