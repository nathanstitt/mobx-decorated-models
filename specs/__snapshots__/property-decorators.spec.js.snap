// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property Decorators can set property to be object 1`] = `
Object {
  "container": undefined,
  "depth": 1,
  "height": 1,
  "id": undefined,
  "metadata": Object {
    "barcode": "Z12",
    "color": "black",
    Symbol(mobx administration): ObservableObjectAdministration$$1 {
      "defaultEnhancer": [Function],
      "keysAtom": Atom$$1 {
        "diffValue": 0,
        "isBeingObserved": false,
        "isPendingUnobservation": false,
        "lastAccessedBy": 0,
        "lowestObserverState": 2,
        "name": "Box@81.metadata.keys",
        "observers": Set {},
      },
      "name": "Box@81.metadata",
      "pendingKeys": Map {
        "asymmetricMatch" => false,
        "nodeType" => false,
        "$$typeof" => false,
        "@@__IMMUTABLE_ITERABLE__@@" => false,
        "@@__IMMUTABLE_RECORD__@@" => false,
        "_isMockFunction" => false,
        "toJSON" => false,
      },
      "proxy": [Circular],
      "target": Object {
        "barcode": "Z12",
        "color": "black",
        Symbol(mobx administration): [Circular],
      },
      "values": Map {
        "barcode" => "Z12",
        "color" => "black",
      },
    },
  },
  "sides": Array [],
  "vessel": undefined,
  "watercraft": undefined,
  "width": 3,
}
`;

exports[`Property Decorators session props are set but not serialized 1`] = `
Object {
  "container": undefined,
  "depth": 1,
  "height": 1,
  "id": undefined,
  "metadata": Object {
    Symbol(mobx administration): ObservableObjectAdministration$$1 {
      "defaultEnhancer": [Function],
      "keysAtom": Atom$$1 {
        "diffValue": 0,
        "isBeingObserved": false,
        "isPendingUnobservation": false,
        "lastAccessedBy": 0,
        "lowestObserverState": -1,
        "name": "ObservableObject@63.keys",
        "observers": Set {},
      },
      "name": "ObservableObject@63",
      "pendingKeys": Map {
        "$$typeof" => false,
        "nodeType" => false,
        "@@__IMMUTABLE_ITERABLE__@@" => false,
        "@@__IMMUTABLE_RECORD__@@" => false,
        "_isMockFunction" => false,
        "toJSON" => false,
      },
      "proxy": [Circular],
      "target": Object {
        Symbol(mobx administration): [Circular],
      },
      "values": Map {},
    },
  },
  "sides": Array [],
  "vessel": undefined,
  "watercraft": undefined,
  "width": 3,
}
`;

exports[`Property Decorators sets an inverse for belongsTo 1`] = `
Object {
  "box": Object {
    "container": undefined,
    "depth": 1,
    "height": 1,
    "id": undefined,
    "metadata": Object {
      Symbol(mobx administration): ObservableObjectAdministration$$1 {
        "defaultEnhancer": [Function],
        "keysAtom": Atom$$1 {
          "diffValue": 0,
          "isBeingObserved": false,
          "isPendingUnobservation": false,
          "lastAccessedBy": 0,
          "lowestObserverState": -1,
          "name": "ObservableObject@59.keys",
          "observers": Set {},
        },
        "name": "ObservableObject@59",
        "pendingKeys": Map {
          "$$typeof" => false,
          "nodeType" => false,
          "@@__IMMUTABLE_ITERABLE__@@" => false,
          "@@__IMMUTABLE_RECORD__@@" => false,
          "_isMockFunction" => false,
          "toJSON" => false,
        },
        "proxy": [Circular],
        "target": Object {
          Symbol(mobx administration): [Circular],
        },
        "values": Map {},
      },
    },
    "sides": Array [],
    "watercraft": undefined,
    "width": 42,
  },
  "cargoCount": undefined,
  "embarks": null,
  "homePorts": Array [],
  "name": "HMS Mobx",
  "radio": null,
  "registration": "",
}
`;
